examples/state_machine/states.py

```diff
@@ -0,0 +1,471 @@
+            events.SubscriptionRestoredEvent.__name__: self.subscription_restored,
+            events.SubscriptionChangedEvent.__name__: self.subscription_changed,
+        }
+
+    def on_enter(self, context: Union[None, PNContext]):
+        self._context.update(context)
+        super().on_enter(self._context)
+        return effects.HandshakeEffect(self._context.channels, self._context.groups)
+
+    def on_exit(self):
+        super().on_exit()
+        return effects.CancelHandshakeEffect()
+
+    def subscription_changed(self, event: events.SubscriptionChangedEvent, context: PNContext) -> PNTransition:
+        self._context.update(context)
+        self._context.channels = event.channels
+        self._context.groups = event.groups
+        self._context.tt = 0
+
+        if event.channels[0] == 'fail':
+            next_state = HandshakeReconnectingState \
+                if context.handshake_count < HandshakingState.MAX_HANDSHAKE_RETRY_COUNT \
+                else HandshakeFailedState
+        else:
+            next_state = HandshakingState
+
+        return PNTransition(
+            state=next_state,
+            context=self._context
+        )
+
+    def subscription_restored(self, event: events.SubscriptionRestoredEvent, context: PNContext) -> PNTransition:
+        self._context.update(context)
+        self._context.channels = event.channels
+        self._context.groups = event.groups
+        self._context.tt = 0
+
+        if event.channels[0] == 'fail':
+            next_state = HandshakeReconnectingState \
+                if context.handshake_count < HandshakingState.MAX_HANDSHAKE_RETRY_COUNT \
+                else HandshakeFailedState
+        else:
+            next_state = ReceivingState
+
+        return PNTransition(
+            state=next_state,
+            context=self._context
+        )
+
+    def reconnecting(self, event: events.HandshakeFailureEvent, context: PNContext) -> PNTransition:
+        self._context.update(context)
+        self._context.handshake_count + 1
+
+        if event.channels[0] == 'fail':
+            next_state = HandshakeReconnectingState \
+                if self._context.handshake_count < HandshakingState.MAX_HANDSHAKE_RETRY_COUNT \
+                else HandshakeFailedState
+        else:
+            next_state = HandshakeReconnectingState
+
+        return PNTransition(
+            state=next_state,
+            context=self._context
+        )
+
+    def disconnect(self, event: events.SubscriptionChangedEvent, context: PNContext) -> PNTransition:
+        self._context.update(context)
+        self._context.channels = event.channels
+        self._context.groups = event.groups
+        self._context.tt = 0
+
+        return PNTransition(
+            state=HandshakeStoppedState,
+            context=self._context
+        )
+
+    def handshaking_success(self, event: events.HandshakeSuccessEvent, context: PNContext) -> PNTransition:
+        return PNTransition(
```

Comment:
We decided that we'll emitStatus on this as well. Look how it looks like in Kotlin:


```
transitionTo(
                        state = Receiving(channels, channelGroups, event.subscriptionCursor),
                        SubscribeEffectInvocation.EmitStatus(
                            PNStatus(
                                category = PNStatusCategory.PNConnectedCategory,
                                operation = PNOperationType.PNSubscribeOperation, // todo is PNSubscribeOperation correct operation
                                error = false,
                                affectedChannels = channels,
                                affectedChannelGroups = channelGroups
                            )
                        )
                    )
```